### PR TITLE
Fix py-dill source archive extension

### PR DESCRIPTION
**Pull request by iarspider**

### CODE REVIEW

File: var/spack/repos/builtin/packages/py-dill/package.py [hunk 1]
```diff
@@ -11,6 +11,7 @@ class PyDill(PythonPackage):
 
     homepage = "https://github.com/uqfoundation/dill"
     url      = "https://pypi.io/packages/source/d/dill/dill-0.2.6.zip"
+    list_url = "https://pypi.io/packages/source/d/dill/"
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-dill/package.py]: This isn't necessary, as it is already the default list_url for most packages.

File: var/spack/repos/builtin/packages/py-dill/package.py [hunk 2]
```diff
@@ -23,3 +24,14 @@ class PyDill(PythonPackage):
...
+
+    def url_for_version(self, version):
+        url = self.list_url
+
+        if version >= Version('0.2.7'):
+            url += 'dill-{0}.tar.gz'
+        else:
+            url += 'dill-{0}.zip'
+
+        url = url.format(str(version))
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-dill/package.py]: No need to convert this to a string, format already does that

File: var/spack/repos/builtin/packages/py-dill/package.py (hunk 1 above)
Comment (iarspider) [reply var/spack/repos/builtin/packages/py-dill/package.py]: It is not. If I remove this line, spack will complain:

```console
$ spack -d install py-dill@0.2.7 ^python@2.7.16
(...)
Traceback (most recent call last):
  File "/build/razumov/spack/bin/spack", line 64, in <module>
    sys.exit(spack.main.main())
  File "/build/razumov/spack/lib/spack/spack/main.py", line 715, in main
    return _invoke_command(command, parser, args, unknown)
  File "/build/razumov/spack/lib/spack/spack/main.py", line 457, in _invoke_command
    return_val = command(parser, args)
  File "/build/razumov/spack/lib/spack/spack/cmd/install.py", line 331, in install
    install_spec(args, kwargs, abstract, concrete)
  File "/build/razumov/spack/lib/spack/spack/cmd/install.py", line 209, in install_spec
    spec.package.do_install(**kwargs)
  File "/build/razumov/spack/lib/spack/spack/package.py", line 1590, in do_install
    partial = self.check_for_unfinished_installation(keep_prefix, restage)
  File "/build/razumov/spack/lib/spack/spack/package.py", line 1854, in check_for_unfinished_installation
    if restage and self.stage.managed_by_spack:
  File "/build/razumov/spack/lib/spack/spack/package.py", line 806, in stage
    self._stage = self._make_stage()
  File "/build/razumov/spack/lib/spack/spack/package.py", line 777, in _make_stage
    composite_fetcher = self.fetcher
  File "/build/razumov/spack/lib/spack/spack/package.py", line 885, in fetcher
    self._fetcher = self._make_fetcher()
  File "/build/razumov/spack/lib/spack/spack/package.py", line 871, in _make_fetcher
    root_fetcher = fs.for_package_version(self, self.version)
  File "/build/razumov/spack/lib/spack/spack/fetch_strategy.py", line 1289, in for_package_version
    return _from_merged_attrs(fetcher, pkg, version)
  File "/build/razumov/spack/lib/spack/spack/fetch_strategy.py", line 1246, in _from_merged_attrs
    url = pkg.url_for_version(version)
  File "/build/razumov/spack/var/spack/repos/builtin/packages/py-dill/package.py", line 34, in url_for_version
    else:
TypeError: unsupported operand type(s) for +=: 'NoneType' and 'str'
```

I can rename this variable into something else and move it inside `url_for_version` if you want.

Comment (adamjstewart) [reply var/spack/repos/builtin/packages/py-dill/package.py]: You could move the variable into `url_for_version` is what I'm saying

Comment (iarspider) [reply var/spack/repos/builtin/packages/py-dill/package.py]: Done.

File: var/spack/repos/builtin/packages/py-dill/package.py [hunk 3]
```diff
@@ -11,7 +11,8 @@ class PyDill(PythonPackage):
 
     homepage = "https://github.com/uqfoundation/dill"
     url      = "https://pypi.io/packages/source/d/dill/dill-0.2.6.zip"
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-dill/package.py]: Can you change this extension to `dill-0.2.7.tar.gz`? That will allow `spack checksum` to pick up new versions in the future.

Comment (iarspider) [reply var/spack/repos/builtin/packages/py-dill/package.py]: Ah, I think I see why it's needed.

**Comment by adamjstewart** [comment]
Flake8:
```
var/spack/repos/builtin/packages/py-dill/package.py:14: [W293] blank line contains whitespace
```

**Pull request merged**